draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -3614,6 +3623,112 @@ Steps:
 38. return (xEn, xEd, yEn, yEd)
 ~~~
 
+## Cofactor Clearing for BLS12-381 G2 {#clear-cofactor-bls12381-g2}
+
+The curve BLS12-381, whose parameters are defined in {{suites-bls12381-g2}},
+admits an efficiently-computable endomorphism psi that can be used to
+speed up cofactor clearing for G2 {{SBCDK09}} {{FKR11}} {{BP17}} (see also
+{{cofactor-clearing}}).
+This section implements the endomorphism psi and a fast cofactor clearing
+method described by Budroni and Pintore {{BP17}}.
+
+The functions in this section operate on points whose coordinates are
+represented as ratios, i.e., (xn, xd, yn, yd) corresponds to the point
+(xn / xd, yn / yd); see {{projective-coords}} for further discussion of
+projective coordinates.
+When points are represented in affine coordinates, one can simply ignore
+the denominators xd == yd == 1.
+
+The following function computes the Frobenius endomorphism for an element
+of F = GF(p^2) with basis (1, I), where I^2 + 1 == 0 in F.
+(This is the base field of the BLS12-381 curve containing the subgroup G2.)
+
+~~~
+frobenius(x)
+
+Input: x, an element of GF(p^2).
+Output: a, an element of GF(p^2).
+
+Notation: x = x0 + I * x1, where x0 and x1 are elements of GF(p).
+
+Steps:
+1. a = x0 - I * x1
+2. return a
+~~~
+
+The following function computes the endomorphism psi for points on the BLS12-381
+curve defined over GF(p^2).
+
+~~~
+psi(xn, xd, yn, yd)
+
+Input: P, the point (xn / xd, yn / yd) on the BLS12-381 curve
+       containing the subgroup G2.
+Output: Q, a point on the same curve.
+
+Constants:
+1. c1 = 1 / (1 + I)^((p - 1) / 3)           # in GF(p^2)
+2. c2 = 1 / (1 + I)^((p - 1) / 2)           # in GF(p^2)
+
+Steps:
+1. qxn = c1 * frobenius(xn)
+2. qxd = frobenius(xd)
+3. qyn = c2 * frobenius(yn)
+4. qyd = frobenius(yd)
+5. return (qxn, qxd, qyn, qyd)
+~~~
+
+The following function computes psi(psi(P)).
+
+~~~
+psi2(xn, xd, yn, yd)
+
+Input: P, the point (xn / xd, yn / yd) on the BLS12-381 curve
+       containing the subgroup G2.
+Output: Q, a point on the same curve.
+
+Constants:
+1. c1 = 1 / 2^((p - 1) / 3)                 # in GF(p^2)
+
+Steps:
+1. qxn = c1 * xn
+2. qyn = -yn
+3. return (qxn, xd, qyn, yd)
+~~~
+
+The following function maps any point on the BLS12-381 curve defined over GF(p^2)
+into the prime-order subgroup G2.
+This function returns a point equal to h\_eff * G2, where h\_eff is the parameter
+given in {{suites-bls12381-g2}}.
+
+~~~
+clear_cofactor_bls12381_g2(P)
+
+Input: P, the point (xn / xd, yn / yd) on the BLS12-381 curve
+       containing the subgroup G2.
+Output: Q, a point in the subgroup G2 of BLS12-381.
+
+Constants:
+1. c1 = -15132376222941642752       # the BLS parameter for BLS12-381
```

Comment:
Can we reference draft-yonezawa-pairing-friendly-curves for this constant? (Oddly, I don't see it in the RG version of that document.)

Reply:
Looks like it's in Appx C of the current RG draft. They demoted BLS12-381, I think...

Reply:
Added a link in the {{suites}} section where we introduce BLS12-381. The appendix backrefs to there. Cool?

Reply:
Yep, works for me!